# Tutorial básico Python - https://github.com/Caldeiraaf 

Este tutorial é uma introdução ao `Python 3`. O objetivo desse tutorial é fornecer um conjunto de habilidades 
básicas  necessárias  para realização das atividades e exercícos das disciplinas Sinais e Sistemas e Controle.
Se você não possui o Jupyter instalado, você pode seguir este tutorial usando um serviço jupyter online:

https://try.jupyter.org/

## Celulas Entrada e Saída

In [1]:
1+2

3

## A função `print` 

**Aviso:** `print` é uma função em  `Python 3`. Você deve usar parênteses em torno do seu parâmetro.

In [2]:
print(2+2)

4


## Variáveis

Existem muitos tipos de variáveis no Python 3. Aqui está uma lista dos tipos mais comuns:
    
### Tipos Numéricos:

- bool (Booleana)
- int (Inteira / longa)
- float (ponto flutuante)
- complex (complexa)

**Aviso:** No `Python 3`, `int` representa  tanto variáveis do tipo inteiro quanto longas (long).

In [3]:
a = 4
b = 1.5
c = 121212121212121212121212121212121212121212121212121212
d = 1j
e = 1/3
f = True

In [4]:
a+b

5.5

In [5]:
a*c

484848484848484848484848484848484848484848484848484848

In [6]:
(b+d)*a

(6+4j)

In [7]:
a+f

5

In [8]:
type(1.5)

float

## Outros tipos de variáveis:

- str (String)
- list (Ordered Array)
- tuple (Ordered Immutable Array)
- dict (Unordered list of keys and values)

In [9]:
my_name = "Andre"
print(my_name)

Andre


In [10]:
my_list = [1,2,3,4,5]
my_list

[1, 2, 3, 4, 5]

In [11]:
my_list + [6]

[1, 2, 3, 4, 5, 6]

In [12]:
my_list

[1, 2, 3, 4, 5]

In [13]:
my_list += [6,7,8]
my_list

[1, 2, 3, 4, 5, 6, 7, 8]

In [14]:
my_list.append(9)
my_list

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [15]:
my_tuple = (1,2,3)
my_tuple

(1, 2, 3)

In [16]:
my_tuple + (4,5,6)

(1, 2, 3, 4, 5, 6)

In [17]:
my_dict = {"name":"Andre", "credit":100}
my_dict

{'name': 'Andre', 'credit': 100}

In [18]:
my_dict["name"]

'Andre'

In [19]:
my_dict["level"] = 4
my_dict

{'name': 'Andre', 'credit': 100, 'level': 4}

In [20]:
my_dict.values()

dict_values(['Andre', 100, 4])

In [21]:
my_dict.keys()

dict_keys(['name', 'credit', 'level'])

## Seleção / Fatiamento (Slicing)

Use a função `len` para medir o comprimento da `list`.

In [22]:
len(my_list)

9

Para acessar um único valor em uma lista, use esta sintaxe:
    
```python
list_name[index]
```

In [23]:
my_list[0]

1

Para selecionar vários valores de uma lista, use esta sintaxe:

```python
index[start:end:step]
```

In [24]:
my_list[1:4]

[2, 3, 4]

In [25]:
my_list[0:6:2]

[1, 3, 5]

In [26]:
my_list[3:]

[4, 5, 6, 7, 8, 9]

**Aviso:** índice negativo seleciona o final da `list`

In [27]:
my_list[-1]

9

In [28]:
my_list[-2]

8

Você pode usar a indexação negativa na seleção de vários valores.

In [30]:
my_list[-2:]

[8, 9]

In [31]:
my_list[:-2]

[1, 2, 3, 4, 5, 6, 7]

In [32]:
my_list[3:-1]

[4, 5, 6, 7, 8]

A terceira localização no índice é o passo (step). 
Se o passo for negativo, a lista será retornada em ordem decrescente.

In [33]:
my_list[::2]

[1, 3, 5, 7, 9]

In [34]:
my_list[3::2]

[4, 6, 8]

In [35]:
my_list[::-1]

[9, 8, 7, 6, 5, 4, 3, 2, 1]

## Trabalhando com Strings

Você pode selecionar uma String como uma `list` utilizando a seguinte sintaxe:

```python
my_string[star:end:step]
```

In [36]:
my_name

'Andre'

In [37]:
my_name[0]

'A'

In [38]:
my_name[0:5:2]

'Ade'

**Aviso:** Você também pode usar a indexação negativa.

In [39]:
my_name[:2]

'An'

## Formatação de String

Você pode usar esta sintaxe para formatar uma String:

```python
some_variable = 50
x = "Value: %s" % some_variable
print(x) # Value: 50
```

In [40]:
first_name = "Andre"
last_name = "Caldeira"
formatted_name = "%s, %s." % (last_name, first_name[0])
print(formatted_name)

Caldeira, A.


Outros formatadores podem ser usados para formatar números:

In [41]:
print("π ≈ %.2f" % 3.14159)

π ≈ 3.14


Para encontrar símbolos unicode:

http://www.fileformat.info/info/unicode/char/search.htm

In [42]:
homeworks = 15.75
midterm = 22
final = 51
total = homeworks + midterm + final

print("Homeworks: %.2f\nMid-term: %.2f\nFinal: %.2f\nTotal: %.2f/100" % (homeworks, midterm, final, total))

Homeworks: 15.75
Mid-term: 22.00
Final: 51.00
Total: 88.75/100


## Usando a função `format(*args, **kwargs)` 

In [43]:
url = "http://{language}.wikipedia.org/"
url = url.format(language="en")
url

'http://en.wikipedia.org/'

In [44]:
## Matemática

In [45]:
1+1

2

In [46]:
4-5

-1

**Aviso:** O comportamento padrão da divisão no Python 3 é a divisão utilizando ponto flutuante (float). 
    Para usar a divisão inteira como Python 2, use `//`

In [47]:
14/5

2.8

In [48]:
14//5

2

In [49]:
2*5

10

Para elevar um número para qualquer potência, use o asterisco `**`. 
Para representar $ a ^ {n} $:

```python
a**n
```

In [50]:
2**3

8

Para calcular o restante (operador de módulo), use `%`. Para representar $ a \mod b = r $:

```python
a % b # Returns r
```

In [51]:
10 % 3

1

Você pode usar a biblioteca `math` para acessar uma variedade de ferramentas para álgebra e geometria. 
Para importar uma biblioteca, você pode usar uma destas sintaxes:

```python
import library_name
import library_name as alias
from module_name import some_class
```

In [52]:
import math

In [53]:
n=52
k=1
math.factorial(n) / (math.factorial(k) * math.factorial(n-k))

52.0

## Loops

In [54]:
for counter in [1,2,3,4,5,6]:
    print(counter)

1
2
3
4
5
6


### Range

No Python 3, `range` é um tipo de comando que gera uma lista de números.

```python
range(stop)
range(start,stop[ ,step])
```

**Aviso:** No Python 2, `range` é uma função que retorna uma lista.
     No Python 3, `range` retorna um iterável do tipo` range`.
     Se você precisar obter uma lista, poderá usar a função `list ()`:

```python
list(range(start,stop[, step]))
```

In [55]:
for counter in range(5):
    print(counter)

0
1
2
3
4


In [56]:
for counter in range(1,5):
    print(counter)

1
2
3
4


In [57]:
for counter in range(2,10,2):
    print(counter)

2
4
6
8


**Aviso:** No Python 3, use `==` para verificar se dois valores são iguais. 
    Para verificar se dois valores não são iguais, use `! =`

## While Loop

In [58]:
counter =1
while counter < 5:
    print(counter)
    counter += 1 # counter = counter+1

1
2
3
4


## If .. Else

In [59]:
if math.pi == 3.2:
    print("Edward J. Goodwin was right!")
else:
    print("π is irrational")

π is irrational


In [60]:
if math.sqrt(2) == (10/7):
    print("Edward J. Goodwin was right!")
elif math.sqrt(2) != (10/7):
    print("Square root of 2 is irrational")

Square root of 2 is irrational


In [61]:
probability = 0.3

if probability >= 0.75:
    print("Sure thing")
elif probability >= 0.5:
    print("Maybe")
elif probability >= 0.25:
    print("Unusual")
else:
    print("No way")

Unusual


## Funções

As funções são definidas no Python usando a palavra-chave `def`.

In [62]:
def get_circumference(r):
    return math.pi * r * 2

In [63]:
get_circumference(5)

31.41592653589793

In [64]:
def binomilal_coef(n,k):
    """
    This function returns the binominal coef
    
    Parameters:
    ===========
    n, k int
    
    return n!/(k!*(n-k)!)
    """
    value = math.factorial(n)/(math.factorial(k)*math.factorial(n-k))
    return value


binomilal_coef(52,2)

1326.0

In [65]:
**Aviso:** Você pode usar SHIFT + TAB para mostrar a sequência de caracteres de uma função.

SyntaxError: invalid syntax (<ipython-input-65-3f23b6a5e909>, line 1)

## Se você gosta de matemática

Esta foi a função binomial que indica quantas maneiras existem para selecionar k número de amostras de um espaço de amostra com n itens se a ordem não importam e as amostras escolhidas não puderem ser repetidas (escolhidas novamente).

Como exemplo, quantas maneiras de selecionar 2 cartas de um baralho de 52 cartas se a ordem da seleção não importa e você não pode escolher a mesma carta duas vezes?
$${n \choose k} = {n! \over {k!(n-k)!}}$$

$${52 \choose 2} = {52! \over {2!(52-2)!}} = {52! \over {2!(50)!}} =  1326$$
